deepchem/models/torch_models/hf_models.py

```diff
@@ -0,0 +1,407 @@
+            (inputs, labels, weights).
+        max_checkpoints_to_keep: int
+            the maximum number of checkpoints to keep.  Older checkpoints are discarded.
+        checkpoint_interval: int
+            the frequency at which to write checkpoints, measured in training steps.
+            Set this to 0 to disable automatic checkpointing.
+        restore: bool
+            if True, restore the model from the most recent checkpoint and continue training
+            from there.  If False, retrain the model from scratch.
+        variables: list of torch.nn.Parameter
+            the variables to train.  If None (the default), all trainable variables in
+            the model are used.
+        loss: function
+            a function of the form f(outputs, labels, weights) that computes the loss
+            for each batch.  If None (the default), the model's standard loss function
+            is used.
+        callbacks: function or list of functions
+            one or more functions of the form f(model, step) that will be invoked after
+            every step.  This can be used to perform validation, logging, etc.
+        all_losses: Optional[List[float]], optional (default None)
+            If specified, all logged losses are appended into this list. Note that
+            you can call `fit()` repeatedly with the same list and losses will
+            continue to be appended.
+
+        Returns
+        -------
+        The average loss over the most recent checkpoint interval
+        """
+        if not isinstance(callbacks, SequenceCollection):
+            callbacks = [callbacks]
+        self._ensure_built()
+        self.model.train()
+        avg_loss = 0.0
+        last_avg_loss = 0.0
+        averaged_batches = 0
+        if variables is None:
+            optimizer = self._pytorch_optimizer
+            lr_schedule = self._lr_schedule
+        else:
+            var_key = tuple(variables)
+            if var_key in self._optimizer_for_vars:
+                optimizer, lr_schedule = self._optimizer_for_vars[var_key]
+            else:
+                optimizer = self.optimizer._create_pytorch_optimizer(variables)
+                if isinstance(self.optimizer.learning_rate,
+                              LearningRateSchedule):
+                    lr_schedule = self.optimizer.learning_rate._create_pytorch_schedule(
+                        optimizer)
+                else:
+                    lr_schedule = None
+                self._optimizer_for_vars[var_key] = (optimizer, lr_schedule)
+        time1 = time.time()
+
+        # Main training loop.
+
+        for batch in generator:
+            if restore:
+                self.restore()
+                restore = False
+            inputs: OneOrMany[torch.Tensor]
+            inputs, labels, weights = self._prepare_batch(batch)
+
+            optimizer.zero_grad()
+            outputs = self.model(**inputs)
```

Comment:
Let's consider whether we should generalize TorchModel instead of reimplementing fit here since a lot of this code is identical to TorchModel

Reply:
There are a couple of ways we can generalize TorchModel here maintaining backward compatability.
1. check type of inputs and if it is an instance of `Dict`, then invoke as `model(**inputs)` else the old style - `model(inputs)`
2. Delegate `model(inputs)` call to a function - say `forward()` which can be implemented in child classes as well.

Reply:
Documenting offline discussion, it's ok to leave this code as-is for now, but we should come back and consider refactoring for maintainability